Issue 4577 - Add GitHub actions

vashirov opened the pull request:
Add first set of actions to compile project using gcc and clang.

Relates: https://github.com/389ds/389-ds-base/issues/4577

droideck commented on .github/workflows/compile.yml:
```diff
@@ -0,0 +1,77 @@
+
+        include:
+          - name: Fedora 33 GCC
+            image: fedora:33
```
I haven't tried yet but can we use Fedora 34 here?
There is image fedora:34 but I haven't checked if there will be any other issues.

In general, I think, it's better to test 389-ds 'master' on the distro for which we build RPMs - master==fedora-rawhide.

vashirov replied:
> I haven't tried yet but can we use Fedora 34 here?
There is image fedora:34 but I haven't checked if there will be any other issues.

We can, but it's not branched yet, it's still rawhide. We expect some unrelated failures due to instability of rawhide. I'd prefer to add F34 once it's branched.

> In general, I think, it's better to test 389-ds 'master' on the distro for which we build RPMs - master==fedora-rawhide.

The point of this test is to use stable tools (gcc, clang) to test unstable code (master). If tools are from an unstable version, we can get unpredictable results. I can add another variant to the matrix, but I'm not sure how useful it would be.

droideck replied:
Okay, sounds good and I think you're right.

The only issues I'm worried about are the Fedora changes related issues (like crypto policies may be different on F33 and rawhide, etc.).
But it's really rare and we can catch it separately.

droideck commented on .github/workflows/compile.yml:
```diff
@@ -0,0 +1,77 @@
+pcre-devel procps-ng psmisc python3-argcomplete python3-argparse-manpage \
+python3-devel python3-ldap python3-libselinux python3-pip \
+python3-policycoreutils python3-pyasn1 python3-pyasn1-modules rpm-build \
+rsync rust systemd-devel uuid wget zlib-devel"
```
I think it will look cleaner line by line because it makes it easier to see the list of dependencies and it's a cleaner git history if we edit any of the dependencies.

I think YAML allows the syntax like this:

```
run: >
  dnf -y install
  389-ds-base
  389-ds-base-snmp
  autoconf
  automake
  ...
```

vashirov replied:
Perhaps it would be better if I put this command in a script? Then it can be reused for other jobs.

droideck replied:
Now I'm thinking, it'll be even better to use 'one source of truth' then...

We have dependencies in 'spec' file, so may be we can get them from there? Like we do in https://www.port389.org/docs/389ds/development/building.html

```
# git clone https://github.com/389ds/389-ds-base.git
# cd 389-ds-base
# sudo dnf install `grep "^BuildRequires" rpm/389-ds-base.spec.in | awk '{print $2}' | sed -e "s/%{python3_pkgversion}/3/"`
```

vashirov replied:
I can't use this on Debian or SUSE. It's also out of sync from Fedora spec file (doesn't have `make` for example), it also pulls in `npm` and `nodejs` that are not needed for this test. So I'd prefer to maintain different lists of dependencies for each distro.

droideck commented on .github/workflows/compile.yml:
```diff
@@ -0,0 +1,77 @@
+name: Build policies
```
Is it a typo? I am not sure I follow which 'policies' are referred here...

vashirov replied:
It refers to build policies (on push, on pull request), but I think I lost another `-name` somewhere to name the actual test.

droideck reviewed (approved):
Okay, thank you for resolving my annoying queries :D
You have my ack. Great job!

vashirov committed:
Issue 4577 - Add GitHub actions

Add first set of actions to compile project using gcc and clang.

Relates: https://github.com/389ds/389-ds-base/issues/4577

Reviewed by: @firstyear, @droideck
